backend/app/user_handler.go

```diff
@@ -0,0 +1,325 @@
+		tokenManager: tokenManager,
+		db:           db,
+		config:       config,
+	}
+}
+
+// RegisterInput struct for data needed when user register
+type RegisterInput struct {
+	Name            string `json:"name" binding:"required" validate:"min=3,max=20"`
+	Email           string `json:"email" binding:"required" validate:"mail"`
+	Password        string `json:"password" binding:"required" validate:"password"`
+	ConfirmPassword string `json:"confirm_password" binding:"required" validate:"password"`
+}
+
+// LoginInput struct for login handler
+type LoginInput struct {
+	Email    string `json:"email" binding:"required" validate:"mail"`
+	Password string `json:"password" binding:"required" validate:"password"`
+}
+
+// RefreshTokenInput struct when user refresh token
+type RefreshTokenInput struct {
+	RefreshToken string `json:"refresh_token" binding:"required"`
+}
+
+// EmailInput struct for user when forgetting password
+type EmailInput struct {
+    Email string `json:"email" validate:"required,email"`
+}
+
+
+// VerifyCodeInput struct takes verification code from user
+type VerifyCodeInput struct {
+	Email string `json:"email" binding:"required"`
+	Code  int    `json:"code" binding:"required"`
+}
+
+// ChangePasswordInput struct for user to change password
+type ChangePasswordInput struct {
+	Email           string `json:"email" binding:"required"`
+	Password        string `json:"password" binding:"required" validate:"password"`
+	ConfirmPassword string `json:"confirm_password" binding:"required" validate:"password"`
+}
+
+// RegisterHandler registers user to the system
+func (h *Handler) RegisterHandler(c *gin.Context) {
+	var request RegisterInput
+
+	// check on request format
+	if err := c.ShouldBindJSON(&request); err != nil {
+		log.Error().Err(err).Send()
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request format"})
+		return
+	}
+
+	// password and confirm password should match
+	if request.Password != request.ConfirmPassword {
+		c.JSON(http.StatusBadRequest, gin.H{"error": "password and confirm password don't match"})
+		return
+	}
+
+	// check if user previously exists
+	_, err := h.db.GetUserByEmail(request.Email)
+	if err == nil {
+		c.JSON(http.StatusConflict, gin.H{"error": "user already registered"})
+		return
+	}
+
+	// hash password
+	salt, err := internal.GenerateSalt()
+	if err != nil {
+		log.Error().Err(err).Msg("Error Generating salt for password")
+		c.JSON(http.StatusInternalServerError, gin.H{"error": "error hashing password"})
+		return
+	}
+	hashed := internal.HashPassword(request.Password, salt)
+
+	isAdmin := internal.IsAdmin(request.Email, h.config.Admins)
+
+	user := models.User{
+		Username: request.Name,
+		Email:    request.Email,
+		Password: hashed,
+		Salt:     salt,
+		Admin:    isAdmin,
+	}
+
+	// create user model in db
+	err = h.db.RegisterUser(&user)
+	if err != nil {
+		log.Error().Err(err).Send()
+		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to register user"})
```

Comment:
generally,  for internal server errors user shouldn't see our server errors only `internal server error` message for the user would be good